1D3V Cartesian Support

prkkumar opened the pull request:
This PR introduces 1D3V Cartesian support to WarpX. We adopted the same strategy as for 2D/RZ/3D, i.e. we introduced new preprocessor macro (WARPX_DIM_1D_Z) and corresponding #ifdef conditions. Main updates are listed below:
- [x] build system
- [x] documentation
- [x] algorithms, algorithms:
       - Field solver (Cartesian Yee, CKC, Nodal)
       - Boundary Conditions (PEC, Silver Muller, Periodic)
       - Field Gather
       - Current/charge deposition (Only Esirkepov Current Deposition is implemented at the moment)
       - Particle pusher
       - Filter (Only BilnearFilter)
       - Moving Window
       - Laser Profiles (Only Gaussian is supported)
- [x] data structures & accessors (e.g., particle positions)
- [x] tests
- [x] CI
- [x] I/O
- [x] GNUMake build system

Following algorithms have **not** been adapted for 1D as part of this PR:
- PSATD, PML, QED, Ionization, Collisions, OPENPMD. 

We have also introduced two CI tests, which are 1D adaptations of Langmuir and laser injection tests. Results of the analysis of these two tests are attached below. Input file corresponding to these tests are 
Test 1. Examples/Tests/Langmuir/inputs_1d_multi_rt (Langmuir Test)
Test 2. Examples/Modules/laser_injection/inputs_1d_rt (Laser Injection)

### **Test 1**
![langmuir_multi_1d_analysis](https://user-images.githubusercontent.com/89051199/140201132-a32dd683-a025-4e15-94ce-1e2342b4af30.png)
### **Test 2 (Ex)**
![plt_Ex](https://user-images.githubusercontent.com/89051199/140201188-39b1db98-900e-44c3-a182-3fc413dce8e2.png)

RemiLehe commented on Docs/source/developers/dimensionality.rst:
```diff
@@ -23,35 +24,37 @@ Defines
+==================  ===========  ===========  ===========  ===========
+``AMREX_SPACEDIM``  ``3``        ``2``        ``1``        ``2``
+``WARPX_DIM_3D``    **defined**  *undefined*  *undefined*  *undefined*
+``WARPX_DIM_1D``    *undefined*  *undefined*  **defined**  *undefined*
```
```suggestion
``WARPX_DIM_1D_Z``  *undefined*  *undefined*  **defined**  *undefined*
```

RemiLehe commented on Source/Make.WarpX:
```diff
@@ -58,14 +58,20 @@ include $(AMREX_HOME)/Tools/GNUMake/Make.defs
 endif
 
+ifeq ($(DIM),1)
+     DEFINES += -WARPX_DIM_1D_Z
```
```suggestion
     DEFINES += -DWARPX_DIM_1D_Z
```

RemiLehe commented on .azure-pipelines.yml:
```diff
@@ -20,6 +20,8 @@ jobs:
   strategy:
     matrix:
+      cartesian1d:
+        WARPX_CI_REGULAR_CARTESIAN_1D: 'TRUE'
```
```suggestion
        WARPX_CI_REGULAR_CARTESIAN_1D: 'TRUE'
        WARPX_CI_PSATD: 'FALSE'
```

RemiLehe commented on GNUmakefile:
```diff
@@ -8,10 +8,11 @@ DEBUG = FALSE
 WARN_ALL = TRUE
 #WARN_ERROR=TRUE
```
Could you revert all the changes that were made to this file?

prkkumar replied:
Done.

RemiLehe commented on Source/Diagnostics/BackTransformedDiagnostic.cpp:
```diff
@@ -640,7 +652,7 @@ BackTransformedDiagnostic (Real zmin_lab, Real zmax_lab, Real v_window_lab,
     for (int i = 0; i < N_slice_snapshots; ++i) {
 
-        IntVect slice_ncells_lab ;
+        //IntVect slice_ncells_lab ;
```
Could you remove this line?
```suggestion
```

prkkumar replied:
Done. Thanks!

RemiLehe commented on Source/Diagnostics/BackTransformedDiagnostic.cpp:
```diff
@@ -1213,7 +1229,11 @@ createLabFrameDirectories() {
+#if ( AMREX_SPACEDIM > 1 )
                                     m_prob_ncells_lab_[0],
+#else
+                                    1,
```
Could you explain this change here?
(I don't quite remember what `m_prob_ncells_lab_[0]` stands for here.)

prkkumar replied:
m_prob_ncells_lab_ is the number of cells in the lab-frame. For 3D , this function passes m_prob_ncells_lab[0], m_prob_ncells_lab[1], m_prob_ncells_lab[2] . For 2D , it passes m_prob_ncells_lab[0], 1, m_prob_ncells_lab[1] and for 1D it passes 1,1, m_prob_ncells_lab[0].

RemiLehe commented on Source/FieldSolver/ElectrostaticSolver.cpp:
```diff
@@ -575,21 +583,29 @@ WarpX::computeE (amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >
 #endif
         for ( MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi )
         {
+#if (AMREX_SPACEDIM > 1)
             const Real inv_dx = 1._rt/dx[0];
+#endif
```
Maybe this is slightly more readable if we include it in the `#ifdef` below
```suggestion
```

prkkumar replied:
Agreed. Changed it

RemiLehe commented on Source/FieldSolver/ElectrostaticSolver.cpp:
```diff
@@ -575,21 +583,29 @@ WarpX::computeE (amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >
             const Real inv_dx = 1._rt/dx[0];
+#endif
 #if (AMREX_SPACEDIM == 3)
             const Real inv_dy = 1._rt/dx[1];
```
```suggestion
            const Real inv_dx = 1._rt/dx[0];
            const Real inv_dy = 1._rt/dx[1];
```

prkkumar replied:
done!

RemiLehe commented on Source/FieldSolver/ElectrostaticSolver.cpp:
```diff
@@ -575,21 +583,29 @@ WarpX::computeE (amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >
             const Real inv_dz = 1._rt/dx[2];
-#else
+#elif (AMREX_SPACEDIM == 2)
             const Real inv_dz = 1._rt/dx[1];
```
```suggestion
            const Real inv_dx = 1._rt/dx[0];
            const Real inv_dz = 1._rt/dx[1];
```

prkkumar replied:
done!

RemiLehe commented on Source/FieldSolver/ElectrostaticSolver.cpp:
```diff
@@ -677,12 +701,16 @@ WarpX::computeB (amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >
 #endif
         for ( MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi )
         {
+#if (AMREX_SPACEDIM > 1)
             const Real inv_dx = 1._rt/dx[0];
+#endif
```
```suggestion
```

prkkumar replied:
Done!

RemiLehe commented on Source/FieldSolver/ElectrostaticSolver.cpp:
```diff
@@ -677,12 +701,16 @@ WarpX::computeB (amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >
             const Real inv_dx = 1._rt/dx[0];
+#endif
 #if (AMREX_SPACEDIM == 3)
             const Real inv_dy = 1._rt/dx[1];
```
```suggestion
            const Real inv_dx = 1._rt/dx[0];
            const Real inv_dy = 1._rt/dx[1];
```

prkkumar replied:
Done!

RemiLehe commented on Source/FieldSolver/ElectrostaticSolver.cpp:
```diff
@@ -677,12 +701,16 @@ WarpX::computeB (amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >
             const Real inv_dz = 1._rt/dx[2];
-#else
+#elif (AMREX_SPACEDIM == 2)
             const Real inv_dz = 1._rt/dx[1];
```
```suggestion
            const Real inv_dx = 1._rt/dx[0];
            const Real inv_dz = 1._rt/dx[1];
```

prkkumar replied:
Done!

RemiLehe commented on Examples/Modules/laser_injection/analysis_1d.py:
```diff
@@ -0,0 +1,176 @@
+w0 = 5.*um
+tt = 10.*fs
+t_c = 24.*fs
+# foc_dist = 13.109*um (not actually used)
```
I think that you could remove this line completely.
```suggestion
```

prkkumar replied:
Agreed. Done!

RemiLehe commented on Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:
```diff
@@ -244,18 +252,27 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
                     Bx(i,j,k) = coef1_z * Bx(i,j,k) + coef2_z * Ey(i,j+1,k);
+#elif WARPX_DIM_1D_Z
+                // At the +z boundary (innermost guard cell)
+                if ( apply_hi_z && ( i==domain_box.bigEnd(1)+1 ) )
```
Am I correct that this should be `bigEnd(0)`?
```suggestion
                if ( apply_hi_z && ( i==domain_box.bigEnd(0)+1 ) )
```

prkkumar replied:
You are right. Changed it

RemiLehe commented on Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:
```diff
@@ -244,18 +252,27 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
+                if ( apply_hi_z && ( i==domain_box.bigEnd(1)+1 ) )
+                    Bx(i,j,k) = coef1_z * Bx(i,j,k) - coef2_z * Ey(i,j,k);
+                // At the -z boundary (innermost guard cell)
+                if ( apply_lo_z && ( i==domain_box.smallEnd(1)-1 ) )
```
```suggestion
                if ( apply_lo_z && ( i==domain_box.smallEnd(0)-1 ) )
```

prkkumar replied:
You are right. Changed it

RemiLehe commented on Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:
```diff
@@ -270,25 +287,36 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
                     By(i,j,k) = coef1_z * By(i,j,k) - coef2_z * Ex(i,j+1,k);
+#elif WARPX_DIM_1D_Z
+                // At the +z boundary (innermost guard cell)
+                if ( apply_hi_z && ( i==domain_box.bigEnd(1)+1 ) )
```
```suggestion
                if ( apply_hi_z && ( i==domain_box.bigEnd(0)+1 ) )
```

prkkumar replied:
You are right. Changed it

RemiLehe commented on Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:
```diff
@@ -270,25 +287,36 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
+                if ( apply_hi_z && ( i==domain_box.bigEnd(1)+1 ) )
+                    By(i,j,k) = coef1_z * By(i,j,k) + coef2_z * Ex(i,j,k);
+                // At the -z boundary (innermost guard cell)
+                if ( apply_lo_z && ( i==domain_box.smallEnd(1)-1 ) )
```
```suggestion
                if ( apply_lo_z && ( i==domain_box.smallEnd(0)-1 ) )
```

prkkumar replied:
You are right. Changed it

RemiLehe commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianCKCAlgorithm.H:
```diff
@@ -252,6 +275,8 @@ struct CartesianCKCAlgorithm {
              + betazx * (F(i+1,j+1,k  ,ncomp) - F(i+1,j  ,k  ,ncomp)
                       +  F(i-1,j+1,k  ,ncomp) - F(i-1,j  ,k  ,ncomp));
+#elif (defined WARPX_DIM_1D_Z)
+        return alphaz * (F(i  ,j+1,k  ,ncomp) - F(i  ,j  ,k  ,ncomp));
```
Am I correct that this should use `i+1` here?
```suggestion
        return alphaz * (F(i+1,j  ,k  ,ncomp) - F(i  ,j  ,k  ,ncomp));
```

prkkumar replied:
Yes, that's correct. Thanks for catching that!

RemiLehe commented on Source/Filter/BilinearFilter.cpp:
```diff
@@ -80,9 +80,16 @@ void BilinearFilter::ComputeStencils(){
     slen.z = 1;
 #endif
+#if (AMREX_SPACEDIM == 1)
+    slen.z = 0;
```
I think that this variable is later used in the loop over cells, when applying the stencil coefficients. Therefore it should be:
```suggestion
    slen.z = 1;
    slen.y = 1;
```

prkkumar replied:
OK. Added these.

RemiLehe commented on Source/Filter/Filter.cpp:
```diff
@@ -153,6 +155,23 @@ void Filter::DoFilter (const Box& tbx,
+        for         (int iz=0; iz < slen_local.z; ++iz){
+            for     (int iy=0; iy < slen_local.y; ++iy){
+                for (int ix=0; ix < slen_local.x; ++ix){
+                    Real sss = sx[ix]*sz[iy];
+                    d += sss*( tmp(i-ix,j,k,scomp+n)
+                              +tmp(i+ix,j,k,scomp+n));
```
I think that the code in 2D and 3D changed recently. Therefore, it should be:
```suggestion
                    Real sss = sz[iy];
                    d += sss*( src_zeropad(i-ix,j,k,scomp+n)
                              +src_zeropad(i+ix,j,k,scomp+n));
```

prkkumar replied:
Got it. Done!

RemiLehe reviewed (changes requested):
Hi Prabhat,
I am about half-way through in the review, but I wanted to already post some suggestions.
In particular, it seems to me that the GPU version of the filter function might not compile in 1D.
Would you have time to look at the suggestions and apply them if they make sense?
Also, could you then try to compile on GPU in 1D, to make sure that the could works in this case too?

prkkumar commented:
> Hi Prabhat, I am about half-way through in the review, but I wanted to already post some suggestions. In particular, it seems to me that the GPU version of the filter function might not compile in 1D. Would you have time to look at the suggestions and apply them if they make sense? Also, could you then try to compile on GPU in 1D, to make sure that the could works in this case too?

Thanks for the review and suggestions, Remi! I have addressed them. I have also fixed the GPU version of the filter function -- now it compiles on GPU as well.

RemiLehe commented on Source/Make.WarpX:
```diff
@@ -58,14 +58,20 @@ include $(AMREX_HOME)/Tools/GNUMake/Make.defs
 endif
 
+ifeq ($(DIM),1)
+     DEFINES += -DWARPX_DIM_1D_Z
```
Could you correct the indentation here?

prkkumar replied:
Yes, fixed it.

RemiLehe commented on Source/Parallelization/WarpXComm_K.H:
```diff
@@ -696,6 +802,7 @@ void warpx_interp (const int j,
 
     else // PSATD (finite-order interpolation)
     {
+#if (AMREX_SPACEDIM > 1) // 1D not implemented for PSATD
```
In the case `AMREX_SPACEDIM = 1`, could you add a runtime error here (e.g. with `amrex::Abort`)?
I am afraid that otherwise we might forget to implement that part of the code if we later implement PSATD 1D, in which case the code might silently skip the interpolation here - which would probably be difficult to detect and debug then.

prkkumar replied:
Good idea. I added runtime error with amrex::Abort for PSATD 1D.

RemiLehe commented on Source/Particles/LaserParticleContainer.cpp:
```diff
@@ -192,7 +201,9 @@ LaserParticleContainer::LaserParticleContainer (AmrCore* amr_core, int ispecies,
         std::vector<Real> windir(3, 0.0);
-#if (AMREX_SPACEDIM==2)
+#if (AMREX_SPACEDIM==1)
+        windir[2+dir] = 1.0;
```
The code is slightly misleading here.
Could instead you use `amrex::ignore_unused` for `dir` and then simply do
```
windir[2] = 1.0;
```

prkkumar replied:
Agreed. Modified it.

RemiLehe commented on Source/Particles/LaserParticleContainer.cpp:
```diff
@@ -278,6 +291,12 @@ LaserParticleContainer::UpdateContinuousInjectionPosition (Real dt)
+        // In 1D, dir=0 corresponds to z
+        // This needs to be converted in order to index `boost_direction`
+        // which has 3 components, for 1D, 2D, and 3D simulations.
+        m_updated_position[2 + dir] -= WarpX::beta_boost *
```
Same comment on `dir` ; could you use `amrex::ignore_unused` here?

prkkumar replied:
Done!

RemiLehe commented on Source/Utils/WarpXMovingWindow.cpp:
```diff
@@ -63,6 +63,11 @@ WarpX::UpdatePlasmaInjectionPosition (Real a_dt)
+            // In 1D, dir=0 corresponds to z
+            // This needs to be converted in order to index `boost_direction`
+            // which has 3 components, for 1D, 2D, and 3D simulations.
+            WarpX::boost_direction[2+dir] * PhysConst::c * a_dt;
```
Could you use `amrex::ignore_unused` for `dir` here?

RemiLehe commented on Source/Utils/WarpXMovingWindow.cpp:
```diff
@@ -365,13 +370,20 @@ WarpX::shiftMF (MultiFab& mf, const Geometry& geom, int num_shift, int dir,
+                      Real x = 0.0;
+                      Real y = 0.0;
+                      Real fac_z = (1.0 - mf_type[0]) * dx[0]*0.5;
+                      Real z = j*dx[0] + real_box.lo(0) + fac_z;
```
Am I correct that this should be
```suggestion
                      Real z = i*dx[0] + real_box.lo(0) + fac_z;
```
?

prkkumar replied:
Oh yes! That should be i. Thank you!!

RemiLehe reviewed (changes requested):
Thanks a lot again for implementing and finalizing this PR! 
I finished reviewing the full PR and have a few additional inline comments/suggestions? Would you have time to incorporate them in the PR?

prkkumar commented:
> Thanks a lot again for implementing and finalizing this PR! I finished reviewing the full PR and have a few additional inline comments/suggestions? Would you have time to incorporate them in the PR?

Thanks for reviewing the full PR! I have incorporated the suggested changes.

ax3l commented on Examples/Tests/Langmuir/analysis_langmuir_multi_1d.py:
```diff
@@ -0,0 +1,114 @@
+#! /usr/bin/env python
+
+# Copyright 2019 Jean-Luc Vay, Maxence Thevenet, Remi Lehe, Prabhat Kumar
```
```suggestion
# Copyright 2019-2021 Jean-Luc Vay, Maxence Thevenet, Remi Lehe, Prabhat Kumar
```

ax3l commented on Examples/Tests/Langmuir/analysis_langmuir_multi_1d.py:
```diff
@@ -0,0 +1,114 @@
+    print("tolerance = {}".format(tolerance))
+    assert( error_rel < tolerance )
+
+test_name = fn[:-9] # Could also be os.path.split(os.getcwd())[1]
```
Please `import os` and use the pattern that is written here as a comment.
Reason: #2385

prkkumar replied:
Done. Thanks!

ax3l commented on GNUmakefile:
```diff
@@ -8,7 +8,7 @@ DEBUG = FALSE
 #WARN_ERROR=TRUE
 
-#DIM     = 2
+#DIM  = 2
```
```suggestion
#DIM = 1
#DIM = 2
```

ax3l commented on Regression/prepare_file_ci.py:
```diff
@@ -119,7 +129,7 @@ def select_tests(blocks, match_string_list, do_test):
 if ci_regular_cartesian_3d:
-    test_blocks = select_tests(test_blocks, ['dim = 2'], False)
+    test_blocks = select_tests(test_blocks, ['dim = 3'], True)
     test_blocks = select_tests(test_blocks, ['USE_RZ=TRUE'], False)
```
This can now be removed, since we check `dim = 3` now :)
```suggestion
```

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianCKCAlgorithm.H:
```diff
@@ -71,6 +71,13 @@ struct CartesianCKCAlgorithm {
         constexpr Real alphay=0._rt;
+#elif defined WARPX_DIM_1D_Z
+        Real const alphaz = inv_dz;
+        // Other coefficients are 0 in 2D Cartesian
```
```suggestion
        // Other coefficients are 0 in 1D Cartesian
```

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianNodalAlgorithm.H:
```diff
@@ -106,9 +117,9 @@ struct CartesianNodalAlgorithm {
         Real const inv_dy = coefs_y[0];
         return 0.5_rt*inv_dy*( F(i,j+1,k,ncomp) - F(i,j-1,k,ncomp) );
-#elif (defined WARPX_DIM_XZ)
+#elif (defined WARPX_DIM_XZ || WARPX_DIM_1D_Z)
```
```suggestion
#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_1D_Z)
```

ax3l commented on Source/Filter/Filter.cpp:
```diff
@@ -153,6 +155,30 @@ void Filter::DoFilter (const Box& tbx,
 
+        dst(i,j,k,dcomp+n) = d;
+    });
+#else
```
```suggestion
#elif (AMREX_SPACEDIM == 1)
```

ax3l commented on Source/Filter/Filter.cpp:
```diff
@@ -153,6 +155,30 @@ void Filter::DoFilter (const Box& tbx,
+
         dst(i,j,k,dcomp+n) = d;
     });
 #endif
```
```suggestion
#else
    amrex::Abort("Filter not implemented for the current geometry!");
#endif
```

ax3l commented on Source/Filter/Filter.cpp:
```diff
@@ -284,11 +314,14 @@ void Filter::DoFilter (const Box& tbx,
+#else
+                                dst(i,j,k,dcomp+n) += sss*(tmp(i-ix,j,k,scomp+n)
+                                                          +tmp(i+ix,j,k,scomp+n));
 #endif
```
```suggestion
#else
    amrex::Abort("Filter not implemented for the current geometry!");
#endif
```

ax3l commented on Source/Filter/Filter.cpp:
```diff
@@ -284,11 +314,14 @@ void Filter::DoFilter (const Box& tbx,
                                                           +tmp(i+ix,j-iy,k,scomp+n)
                                                           +tmp(i-ix,j+iy,k,scomp+n)
                                                           +tmp(i+ix,j+iy,k,scomp+n));
+#else
```
```suggestion
#elif (AMREX_SPACEDIM == 1)
```

ax3l commented on Source/Filter/NCIGodfreyFilter.cpp:
```diff
@@ -129,3 +130,4 @@ void NCIGodfreyFilter::ComputeStencils(){
 
     Gpu::synchronize();
 }
+#endif //1D
```
```suggestion
```

ax3l commented on Source/Filter/NCIGodfreyFilter.cpp:
```diff
@@ -21,6 +21,7 @@
 #include <vector>
 
 using namespace amrex;
+#if (AMREX_SPACEDIM != 1) //1D not implemented
```
Avoid linker issues:
```suggestion
```

ax3l commented on Source/Filter/NCIGodfreyFilter.cpp:
```diff
@@ -21,6 +21,7 @@
+#if (AMREX_SPACEDIM != 1) //1D not implemented
 
 NCIGodfreyFilter::NCIGodfreyFilter(godfrey_coeff_set coeff_set, amrex::Real cdtodz, bool nodal_gather){
     // Store parameters into class data members
```
Add explicit `#elif` (2D) and `#else` (all other) branches in the body of the function.
Then `amrex::Abort()` with a message for the `#else` branch :)

ax3l commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -282,7 +293,10 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
         // For 2D, only two are expected. The third is overwritten with 1.
-#if WARPX_DIM_XZ
+        // For 1D, only one is expected. The second and third are overwritten with 1.
+#if WARPX_DIM_1D_Z
```
```suggestion
#if defined(WARPX_DIM_1D_Z)
```

ax3l commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -282,7 +293,10 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
+        // For 1D, only one is expected. The second and third are overwritten with 1.
+#if WARPX_DIM_1D_Z
+        constexpr int num_required_ppc_each_dim = 1;
+#elif WARPX_DIM_XZ
```
```suggestion
#elif defined(WARPX_DIM_XZ)
```

ax3l commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -292,6 +306,10 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
         num_particles_per_cell_each_dim.push_back(1);
 #endif
+#if WARPX_DIM_1D_Z
+        num_particles_per_cell_each_dim.push_back(1);
```
I am confused about this, why twice the push_back? If correct, maybe add a comment inline? :)

prkkumar replied:
In 1D, only one number is required for num_particles_per_cell_each_dim. If three are provided, 2nd and third is overwritten by 1. I added a comment inline. :)

ax3l commented:
Awesome, I think this is nearly ready. I added my suggestions inline as well :)

ax3l reviewed (approved):
Fantastic, thanks a lot for this PR!

All green, two reviews, in we go :rocket: :sparkles:
